src/portable/synopsys/dwc2/dcd_dwc2.c

```diff
@@ -630,6 +630,38 @@ void dcd_sof_enable(uint8_t rhport, bool en)
 /* DCD Endpoint port
  *------------------------------------------------------------------*/
 
+#if TU_CHECK_MCU(OPT_MCU_ESP32S2, OPT_MCU_ESP32S3, OPT_MCU_ESP32P4)
+
+// Check if IN/OUT endpoint is avaliable before opening it
+static bool dcd_ep_available(uint8_t rhport, uint8_t dir) 
+{
+    // Verify that we have a vacant EP
+    if ((dwc_ep_config[rhport].in_ep + dwc_ep_config[rhport].out_ep) > (dwc_ep_config[rhport].ep_max_count - 1)) {
+        TU_LOG(1, "Trying to open an endpoint, but max number of endpoints: %d already opened on this target \r\n", dwc_ep_config[rhport].ep_max_count);
+        return false;
+    }
+    // Get the ep_count amount at the moment as a temporal variable and update it
+    uint8_t new_ep_count = (dir) ? dwc_ep_config[rhport].in_ep : dwc_ep_config[rhport].out_ep;
+    new_ep_count++;
+
+    // Verify overflow for IN EP ESP32Sx
+ #if TU_CHECK_MCU(OPT_MCU_ESP32S2, OPT_MCU_ESP32S3)
+     // ESP32Sx has 6 endpoints, from which only 5 can be confiugred as IN
+     if ((dir) && (new_ep_count > (dwc_ep_config[rhport].ep_max_count - 1))) {
+         TU_LOG(1, "Trying to open IN endpoint, but max number of IN endpoints: %d already opened on this target \r\n", dwc_ep_config[rhport].ep_max_count - 1);
+         return false;
+     }
+ #endif // TU_CHECK_MCU(OPT_MCU_ESP32S2, OPT_MCU_ESP32S3)
+    // Write new value back
+    if(dir) {
+        dwc_ep_config[rhport].in_ep = new_ep_count;
+    } else {
+        dwc_ep_config[rhport].out_ep = new_ep_count;
+    }
+    return true; 
+}
+#endif // TU_CHECK_MCU(OPT_MCU_ESP32S2, OPT_MCU_ESP32S3, OPT_MCU_ESP32P4)
```

Comment:
In this block of code we can add two logically connected functions: 
1. One is increase the amount of EPs  (`dcd_ep_available()` already implemented)
2. Another one is decrease the amount of EPs (with asserts to eliminate the wrong code logic from higher layers). 

So, it will be: 
```suggestion

// Decreases counter for opened endpoints (not for EP0)
static void dcd_ep_release(uint8_t rhport, uint8_t dir)
{
  if (dir) {
    TU_ASSERT(dwc_ep_config[rhport].in_ep == 0);
    dwc_ep_config[rhport].in_ep--;
  } else {
    TU_ASSERT(dwc_ep_config[rhport].out_ep == 0);
    dwc_ep_config[rhport].out_ep--;
  }
}
#endif // TU_CHECK_MCU(OPT_MCU_ESP32S2, OPT_MCU_ESP32S3, OPT_MCU_ESP32P4)
```